lib/src/services/song_recognition.dart

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2022, Tomás Sasovsky
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:developer';
+import 'dart:io';
+
+import 'package:crypto/crypto.dart';
+import 'package:http/http.dart' as http;
+import 'package:radio_garden/radio_garden.dart';
+
+import 'package:uuid/uuid.dart';
+
+class SongRecognitionService {
+  SongRecognitionService._() {
+    _httpClient = http.Client();
+  }
+
+  static void init() {
+    _instance = SongRecognitionService._();
+  }
+
+  static SongRecognitionService get instance =>
+      _instance ??
+      (throw Exception('Song Recognition service must be initialised'));
+  static SongRecognitionService? _instance;
+
+  static final SongRecognitionOptions _options = SongRecognitionOptions(
+    accessKey: arcCloudAccessKey,
+    accessSecret: arcCloudAccessSecret,
+    host: arcCloudHost,
+  );
+
+  Uuid get uuid => const Uuid();
+
+  http.Client get httpClient =>
+      _httpClient ?? (throw Exception('Http Client must be initialized'));
+
+  // HttpClient used to get the sample
+  http.Client? _httpClient;
+
+  RadioGardenSearchResponse? _currentRadio;
+
+  RadioGardenSearchResponse get currentRadio =>
+      _currentRadio ??
+      (throw Exception('RadioGarden is getting to know the radio :)'));
+
+  set currentRadio(RadioGardenSearchResponse radio) => _currentRadio = radio;
+
+  /// Generates a sample file used to identify the song playing.
+  ///
+  /// Returns a [File] stored in the /tmp directory.
+  Future<File> generateSample({
+    required String radioId,
+    int durationInSeconds = 10,
+  }) async {
+    final completer = Completer<File>();
+
+    final url =
+        'https://radio.garden/api/ara/content/listen/$radioId/channel.mp3';
+    final uri = Uri.parse(url);
+
+    final request = http.Request('GET', uri);
+    try {
+      final response = await httpClient.send(request);
+
+      StreamSubscription<List<int>>? streamSubscription;
+
+      final bitRate = int.parse(response.headers['icy-br'] ?? '128');
+      final outputFile = File('${Directory.systemTemp.path}/${uuid.v4()}');
+      if (!outputFile.existsSync()) outputFile.createSync();
+      final sink = outputFile.openWrite();
+
+      streamSubscription = response.stream.listen((chunk) async {
+        final bytes = await outputFile.length();
+
+        if (bytes > 0) {
+          // we can calculate the duration by using the bitrate and
+          //the file size, the formula is found here:
+          // http://www.audiomountain.com/tech/audio-file-size.html
+          final bytePerSecond = bitRate / 8 * 1000;
+          final expectedBytes = bytePerSecond * durationInSeconds;
```

Comment:
These should probably go outside of the ´listen´ function, as they get computed needlessly every time we get new chunks.